ops/ruby/devops-toolkit/tasks/logs_based_metrics.rb

```diff
@@ -0,0 +1,68 @@
+require 'google/cloud/logging'
+require_relative 'lib/gcp_environment_visitor.rb'
+
+# Take a source environment and logs-based metric URI and replicate it across all target environments
+# Note well: if two environments share a GCP project, this means they share a metrics namespace.
+# If those projects share a log file (distinguished by MonitoredResource), then we dont' want to duplicate
+# a metric into that shared environment.
+class LogsBasedMetrics
+  def initialize(options)
+    @envs_file = options[:'envs-file']
+    @source_dashboard_number = options[:'source-uri']
+    @source_env_short_name = options[:'source-env']
+    @logger = options[:logger] || Logger.new(STDOUT)
+    @is_dry_run = options[:'replicate']
+    @visitor = GcpEnvironmentVisitor.new(@envs_file, @logger)
+  end
+
+  def replicate
+    # Visit the source environment and smuggle out a metric definition
+    source_metric = get_source_metric
+
+    # Replicate to the other environments (all but source env)
+    copy_to_target_envs(source_metric)
+  end
+
+  private
+
+  # Fetch the source metric by name from the source environment
+  def get_source_metric
+    source_env = @visitor.env_by_short_name(@source_env_short_name)
+    @logger.info("Source environment is #{source_env}")
+    source_metric = nil # extract from its environment
+    @visitor.visit(source_env) do |env|
+      logging_client = Google::Cloud::Logging.new({project: env.project_id})
+      metrics = logging_client.metrics
+      source_metric = metrics.select { |m| m.name == @source_dashboard_number }.first
```

Comment:
how does a metric's name relate to a dashboard number?

Reply:
They're completely independent.  A metric might be represented on zero or dozens of dashboards.

Google just assigns dashboard numbers randomly as far as I can tell in the console, but I have to pick them randomly.